nifi-nar-bundles/nifi-geohash-bundle/nifi-geohash-processors/src/test/java/org/apache/nifi/processors/geohash/GeohashRecordTest.java

```diff
@@ -0,0 +1,151 @@
+import org.apache.commons.io.IOUtils;
+import org.apache.nifi.avro.AvroTypeUtil;
+import org.apache.nifi.controller.ControllerService;
+import org.apache.nifi.json.JsonRecordSetWriter;
+import org.apache.nifi.json.JsonTreeReader;
+import org.apache.nifi.reporting.InitializationException;
+import org.apache.nifi.schema.access.SchemaAccessUtils;
+import org.apache.nifi.serialization.record.MockSchemaRegistry;
+import org.apache.nifi.serialization.record.RecordSchema;
+import org.apache.nifi.util.MockFlowFile;
+import org.apache.nifi.util.TestRunner;
+import org.apache.nifi.util.TestRunners;
+import org.codehaus.jackson.map.ObjectMapper;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class GeohashRecordTest {
+
+    private final TestRunner runner = TestRunners.newTestRunner(GeohashRecord.class);
+
+    @BeforeEach
+    public void setUp() throws InitializationException {
+        ControllerService reader = new JsonTreeReader();
+        ControllerService writer = new JsonRecordSetWriter();
+        ControllerService registry = new MockSchemaRegistry();
+        runner.addControllerService("reader", reader);
+        runner.addControllerService("writer", writer);
+        runner.addControllerService("registry", registry);
+
+        try (InputStream is = getClass().getResourceAsStream("/record_schema.avsc")) {
+            String raw = IOUtils.toString(is, "UTF-8");
+            RecordSchema parsed = AvroTypeUtil.createSchema(new Schema.Parser().parse(raw));
+            ((MockSchemaRegistry) registry).addSchema("record", parsed);
+
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
+        }
+
+        runner.setProperty(reader, SchemaAccessUtils.SCHEMA_REGISTRY, "registry");
+        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_NAME_PROPERTY);
+        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_REGISTRY, "registry");
+
+        runner.setProperty(GeohashRecord.RECORD_READER, "reader");
+        runner.setProperty(GeohashRecord.RECORD_WRITER, "writer");
+        runner.enableControllerService(registry);
+        runner.enableControllerService(reader);
+        runner.enableControllerService(writer);
+
+        runner.setProperty(GeohashRecord.LATITUDE_RECORD_PATH, "/latitude");
+        runner.setProperty(GeohashRecord.LONGITUDE_RECORD_PATH, "/longitude");
+        runner.setProperty(GeohashRecord.GEOHASH_RECORD_PATH, "/geohash");
+        runner.setProperty(GeohashRecord.GEOHASH_FORMAT, GeohashRecord.GeohashFormat.BASE_32.toString());
+        runner.setProperty(GeohashRecord.GEOHASH_LEVEL, "12");
+    }
+
+    private void commonTest(String path, int failure, int success, int original) {
```

Comment:
For clarity of implementation, recommend renaming this to something like `run()` or `assertTransfers()`